### PR TITLE
fix(augmentation): draw a two-value RandomAffine3D scale once per sample

### LINKED ISSUE #4704: RandomAffine3D: scale=(a, b) is documented as isotropic but draws the three axes independently
## Summary

`RandomAffine3D` and `AffineGenerator3D` document a two-value `scale` as isotropic:

> scale: scaling factor interval. If (a, b) represents isotropic scaling, the scale is randomly sampled from the range a <= scale <= b. If ((a, b), (c, d), (e, f)), the scale is randomly sampled from the range a <= scale_x <= b, c <= scale_y <= d, e <= scale_z <= f.

`(a, b)` actually draws the three axes **independently**, so a user asking for aspect-preserving scaling gets an anisotropic one. The 2D `RandomAffine` does the documented thing.

```python
import torch, kornia.augmentation as K
for cls, shape in ((K.RandomAffine3D, (2000, 1, 5, 6, 7)), (K.RandomAffine, (2000, 1, 6, 7))):
    s = cls(0.0, scale=(0.5, 2.0), p=1.0).forward_parameters(torch.Size(shape))["scale"]
    print(cls.__name__, int((s.max(1).values == s.min(1).values).sum()), "/ 2000 rows have one scale for every axis")
# RandomAffine3D 0 / 2000      (largest per-row spread 1.48 of the 1.5 available)
# RandomAffine   2000 / 2000

torch.manual_seed(0); aug = K.RandomAffine3D(0.0, scale=(0.5, 2.0), p=1.0); aug(torch.rand(1, 1, 5, 6, 7))
print(aug.transform_matrix[0].diagonal())   # tensor([1.0787, 1.2447, 1.3457, 1.0000])
```

Measured on `main` @ ff204fb9c, torch 2.14.0, CPU float32.

## Cause

`kornia/augmentation/random_generator/_3d/affine.py`: a shape-`(2,)` `scale` is expanded with `_scale.unsqueeze(0).repeat(3, 1)` into three identical **ranges**, and `forward` then draws from `scale_1_sampler`, `scale_2_sampler` and `scale_3_sampler` separately. The 2D generator draws once and repeats the **value** for a two-value range.

## Expected

For `scale=(a, b)`, draw one value per sample and use it on all three axes, as the docstring says and as `RandomAffine` does. `((a, b), (c, d), (e, f))` stays per-axis. This changes the sampled parameters for a given seed, so it needs a changelog entry.

Found while reviewing #4634 (batch 6d of #4407), whose `RandomAffine3D` Convention block now states the current behaviour and cites this issue.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by thegoodengineer**
## What does this pull request do?

`RandomAffine3D` and `AffineGenerator3D` document a two-value `scale=(a, b)` as isotropic, but the generator expanded it into three identical *ranges* and drew each axis from its own sampler. A caller asking for aspect-preserving scaling got an anisotropic one; the 2D `RandomAffine` does the documented thing.

The issue's reproduction, 2000 rows with `scale=(0.5, 2.0)`, seed 0, before and after:

```text
                                       before (main)                     after (this branch)
RandomAffine3D rows with one scale     0 / 2000 (largest spread 1.4874)  2000 / 2000 (spread 0.0)
RandomAffine   rows with one scale     2000 / 2000                       2000 / 2000
transform_matrix[0].diagonal()         [1.0787, 1.2447, 1.3457, 1.0]     [1.0787, 1.0787, 1.0787, 1.0]
three-pair form rows with one scale    0 / 2000                          0 / 2000 (unchanged)
```

`AffineGenerator3D.make_samplers` now keeps a single `scale_sampler` for the shape-`(2,)` form, and `forward` draws `(batch_size,)` once and repeats it across the three axes, the same `unsqueeze(1).repeat(1, n)` the 2D generator uses. The three-pair form `((a, b), (c, d), (e, f))` keeps its three per-axis samplers and is untouched; the range validation of the two-value form runs once under the name `scale` instead of three times under `scale-x/y/z`.

This changes the random stream of the two-value form for a fixed seed: the scale consumes one draw where it used to consume three, so the translations and shears drawn after it move too. The three-pair form and `scale=None` are bitwise unchanged. It carries a `fixed` changelog fragment.

## Related issue or discussion

Closes #4704.

## Pins moved with the fix

- `test_wart_random_affine3d_two_value_scale_is_not_isotropic_4704` becomes `test_convention_random_affine3d_two_value_scale_is_isotropic_4704`. It keeps the 2D comparison, asserts the three columns are equal and inside `[a, b]` while still varying across the batch, checks the transform-matrix diagonal against the parameters, checks that `same_on_batch=True` gives one value for the whole batch, and checks that the three-pair form still draws each axis independently.
- The Convention bullet on `RandomAffine3D` now states the isotropic rule and the per-axis three-pair form, and no longer cites #4704. `grep -rnE "#4704|issues/4704" kornia/ tests/` is empty apart from the flipped pin's name.
- `TestRandomAffineGen3D::test_random_gen` and `test_same_on_batch` pin seeded literals with the three-pair form, so they are unchanged and still pass. `test_valid_param_combinations` already covers `torch.tensor([0.7, 1.2])` for `B = 0, 1, 8` and both `same_on_batch` values.

## What did you check?

**Nothing outside the two-value form changes.** I captured 144 `AffineGenerator3D` outputs on `main` and on this branch: `scale` `None`, `((0.7, 1.2),) * 3`, `((0.5, 2.0), (0.7, 1.2), (0.9, 1.1))` and `(0.7, 1.2)`; batch shapes `(0, 1, 4, 5, 6)`, `(1, 1, 4, 5, 6)` and `(8, 3, 12, 16, 19)`; RNG dtypes float32 and float64; both `same_on_batch` values; three seeds. Each probe ran from its own tree's root with `kornia.__file__` printed.

```text
scale=None or three-pair                        108 / 108 bitwise identical
scale=(0.7, 1.2), B = 0                          12 /  12 bitwise identical (no draw consumed)
scale=(0.7, 1.2), B > 0                          24 /  24 angles identical (drawn before the scale),
                                                          x-axis scale identical (the first scale draw),
                                                          y/z scale, translations and shears move
```

**The moved pin fails on `main`,** run from a worktree of `main` with the updated test file: `assert (False)` on the column-equality check.

**Compile.** `AffineGenerator3D.forward` traces with `fullgraph=True` on the eager backend for both scale forms and both `same_on_batch` values, and the compiled parameters are `torch.equal` to eager. This machine has no MSVC, so `inductor` cannot build here; CI's dynamo leg is the one to confirm it.

```text
$ pytest tests/augmentation/test_conventions_3d.py tests/augmentation/test_random_generator_3d.py \
      tests/augmentation/test_backward_3d.py tests/augmentation/test_parameter_constants.py \
      -k "affine or Affine" --dtype=float32,float64
181 passed, 4098 skipped (test_backward_3d is marked "To be deprecated")

$ ... --dtype=float16  --xfail-known-failures --known-failure-profile=cpu-float16
92 passed, 2 xfailed

$ ... --dtype=bfloat16 --xfail-known-failures --known-failure-profile=cpu-bfloat16
93 passed, 1 xfailed

$ pytest tests/augmentation/test_conventions_3d.py tests/augmentation/test_random_generator_3d.py \
      tests/augmentation/test_parameter_constants.py tests/augmentation/test_augmentation.py \
      tests/augmentation/container -k "3D or 3d or affine or Affine or parameter" --dtype=float32,float64
658 passed, 4 skipped, 2 xfailed

$ pytest --doctest-modules kornia/augmentation/_3d/geometric/affine.py kornia/augmentation/random_generator/_3d/affine.py
1 passed

$ pre-commit run --files <the three changed files>
all hooks passed
```

I did not run CUDA or MPS locally.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
## Review at `0964c00d8`

The fix is correct and does what #4704 describes. The issue matches the defect. The 2D `AffineGenerator` draws `scale_2_sampler` once and uses `.unsqueeze(1).repeat(1, 2)`, and the 3D two-value branch now does the same with `repeat(1, 3)`. `Closes #4704` is accurate: the closing grep `grep -rnE "#4704|issues/4704" kornia/ tests/ docs/` on the merge tree finds only the renamed pin. The Convention bullet and the `scale` docstrings in both files state the isotropic rule, and the `fixed` fragment documents the seeded-stream change, as in the #4610 precedent. One item blocks the merge: the validation line you moved can be deleted and every test still passes. The fix is below.

### Blocks

**1. Deleting the two-value `scale` range check leaves the whole augmentation test directory green.** The fix moves the bounds check for `(a, b)` onto a new line, `_singular_range_check(_scale, "scale", ...)`. If I replace that line with `pass`, `tests/augmentation` still reports `3932 passed` (float32). A negative isotropic scale such as `(-0.5, 2.0)` would then build a sampler without an error. `test_invalid_param_combinations` has no invalid-scale row for the two-value form. This fix adds a pin that passes on this head and fails under that mutant:

```diff
@@ -491,6 +491,8 @@ class Test3DAugmentationConventions(BaseTester):
         axes = per_axis.forward_parameters(torch.Size([64, 1, 5, 6, 7]))["scale"]
         assert axes.shape == (64, 3)
         assert not bool((axes[:, 0] == axes[:, 1]).any()) and not bool((axes[:, 1] == axes[:, 2]).any())
+        with pytest.raises(ValueError, match="scale out of bounds"):
+            K.RandomAffine3D(0.0, scale=(-0.5, 2.0), p=1.0)
```

The `match` also pins the new error name. `main` reported the same input as `scale-x out of bounds`.

### What I verified

- **Worktrees.** The PR worktree HEAD equals `refs/pull/4731/head`. I also built a `main` (8b03a8085) worktree and a PR+main merge worktree. Each probe printed `kornia.__file__` inside its own tree.
- **Stream comparison, 360 `AffineGenerator3D` configurations per tree.** The grid covered scale `None`, `((0.5, 2.0), (0.7, 1.2), (0.9, 1.1))`, `((0.7, 1.2),) * 3`, `(0.7, 1.2)` and `torch.tensor([0.5, 2.0])`. It crossed each scale with 3 shear forms (including six-value shears), with and without translation, both `same_on_batch` values, B in {0, 1, 7}, and float32/float64. The `None` and three-pair forms, 216/216, are bitwise identical to `main`, including the global RNG state afterwards. In the two-value form, every row has sx == sy == sz inside `[a, b]`. `same_on_batch=True` gives a single value for the whole batch, and the x column equals `main`'s first draw. Only the y/z scales and the translations and shears drawn after them move, as the fragment says.
- **Validation behaviour.** On both trees, `(-1, 2)` and `(1, 2, 3)` raise `ValueError`, and `(1, 1)`, `(0, 1)` and a reversed `(2, 1)` are accepted. The only difference is the error name.
- **The pin fails on `main`.** I ran the PR's test file against the `main` worktree, and `test_convention_random_affine3d_two_value_scale_is_isotropic_4704` fails with `assert (False)`.
- **Mutants against `test_conventions_3d.py` and `test_random_generator_3d.py` (454 passed at baseline):**
  - Killed: three independent `scale_sampler` draws (per axis); two tied axes plus one independent; `same_on_batch` dropped from the isotropic draw; a degenerate `(a, a)` sampler; a shifted `(b, 2b)` range.
  - Survived: only the dropped range check (Block 1).
- **Merge tree.** `pytest tests/augmentation --dtype=float32,float64` gives 7155 passed, 42 xfailed, 7 xpassed (non-strict), 0 failed.
- **CI** is green at this head, including the dynamo, half-precision, MPS and torch 2.5.1 legs.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by thegoodengineer** [comment]
Fixed, new head `5051328f`.

Block 1: added the suggested pin to `test_convention_random_affine3d_two_value_scale_is_isotropic_4704`. `K.RandomAffine3D(0.0, scale=(-0.5, 2.0), p=1.0)` must raise `ValueError` matching `scale out of bounds`, which also pins the new error name (`main` reported it as `scale-x`).

Checked: the pin passes on this head and fails with `DID NOT RAISE ValueError` when the range check is replaced by `pass`. `tests/augmentation/test_conventions_3d.py` and `test_random_generator_3d.py -k "affine or Affine or 4704"` give 179 passed in float32 and float64; pre-commit passes. No source change in this commit.

**Review by ducha-aiki — APPROVED** [review]
## Re-review at `5051328f6`

Block 1 from the previous review is resolved. The new commit adds only the three-line pin to `test_convention_random_affine3d_two_value_scale_is_isotropic_4704`. I re-read the whole PR diff: the generator, the Convention bullet, the fragment and the test. Nothing else changed since `0964c00d8`, and I have no open items. Approving.

### What I verified

- **Worktree.** The PR worktree HEAD is `5051328f62`, the same as `headRefOid`. Each probe printed `kornia.__file__` from its own tree.
- **Mutants.** I ran all six mutants from the previous review again, plus three extra ones, against `test_conventions_3d.py` and `test_random_generator_3d.py` with `--dtype=float32,float64`. The baseline is 454 passed and 2 xfailed. All nine mutants are killed:
  - three independent `scale_sampler` draws: 1 failed
  - two tied axes plus one independent: 1 failed
  - `same_on_batch` dropped from the isotropic draw: 1 failed
  - a degenerate `(a, a)` sampler: 1 failed
  - a shifted `(b, 2b)` range: 1 failed
  - **the range check replaced by `pass`: 1 failed.** This was Block 1. The failure is the new pin, `Failed: DID NOT RAISE <class 'ValueError'>`.
  - `and` changed to `or` in the `forward` branch condition: 53 failed
  - `_isotropic_scale` left `False`: 49 failed
  - the lower bound `0` changed to `-inf` in the two-value check: 1 failed
- **Error-name change.** `(-0.5, 2.0)` still raises `ValueError`. The message now reads `scale out of bounds...` where `main` printed `scale-x out of bounds...`. The three-pair form still reports `scale-x` and `scale-z` by axis, as before. `grep -rnE "scale-x|scale-y|scale-z|scale out of bounds" kornia/ tests/ docs/ changelog.d/ CHANGELOG.md` finds only the generator's three-pair lines and the new pin, so nothing else depends on the old text. The exception type is unchanged, and the new name matches the 2D generator's. This needs no fragment change.
- **Fragment.** `changelog.d/4731.fixed.md` has not changed since the last review, and every statement in it was verified there by the 360-configuration stream comparison. The closing grep, `grep -rnE "#4704|issues/4704" kornia/ tests/ docs/ changelog.d/`, returns nothing on this head.
- **Merge tree with `main` (`4b13aae42`).** It merges cleanly (`merge-tree` rc=0), and merged-vs-`main` is exactly the PR diff (4 files, +34/-13). On the merged tree, `pytest tests/augmentation --dtype=float32,float64` gives 7203 passed, 40 xfailed, 7 xpassed (non-strict) and 0 failed.
- **CI.** All checks pass at `5051328f6`: 39/39 check runs succeeded, and `gh pr checks` lists 41 pass.

*Posted on behalf of @ducha-aiki by Claude (Opus 5.5).*
